docs/RFC/distributed-dns/dnsrecord-lifecycle.md

```diff
@@ -0,0 +1,54 @@
+
+# The idea
+We now will constantly reconcile DNS records. The reasoning is that other controllers may override/change records in the DNS provider so there is a need to requeue the DNS Record from time to time even when no local changes are introduced.
+
+
+# Details
+There are a few new fields on the DNS Record status:
+* QueuedAt is a time when the DNS record was received for the reconciliation
+* QueuedFor is a time when we expect a DNS record to be reconciled again
+* ValidFor indicates the duration since the last reconciliation we consider data in the record to be valid
+* WriteCounter represents a number of consecutive write attempts on the same generation of the record. It is being reset to 0 when the generation changes or there are no changes to write.
+
+
+There is an option to override the `ValidFor` and `DefaultRequeueTime` with `valid-for` and `requeue-time` flags respectively.
+
+
+The `DefaultRequeueTime` is the duration between successful validation and the next reconciliation to ensure that the record is still up-to-date.
+
+
+The `ValidFor` is used to determine if we should do a full reconciliation when we get the record. If the record is still valid we will only update finalizers and validate the record itself. It will not perform anything that involves a DNS provider.
+
+
+## DNS Record normal lifecycle
+Once we enqueue the DNS record, controller will compile a list of changes to the DNS provider and will apply it. After this, the record is enqueued with the `validationRequeueTime` and the `Ready` condition will be marked as `false` with a message `Awaiting Validation`. When the record is received again and the controller ensures there are no changes needed (the ones applied are present in the DNS Provider) it sets the `Ready` condition to `true` and enqueues it with the `defaultRequeueTime`.
+
+
+At any time when the record is requeued we also set the `record.Status.QueuedFor` field with a timestamp for when we expect to receive the record again. And on every reconciliation we set the `record.Status.QueuedAt` to be the time of the reconciliation.
+
+
+Upon deletion, the process will be similar. The controller will determine the changes needed to the DNS provider and will apply them. The record will be requeued with the `validationRequeueTime`. Once we receive it back and ensure that there are no changes needed for the DNS provider we remove the finalizer from the record.
+
+
+The `validationRequeueTime` duration is randomized +/- 50%.
+
+
+## When things go south
+If the record is received prematurely - the `ValidFor` + `QueuedAt` is more than the current time - we requeue it again for the `ValidFor` duration.
+
+
+When we encounter an error during the reconciliation we will not requeue the record and will put in an appropriate error message in the log and on the record. In order for it to reconcile again there must be a change to the DNS Record CR.
```

Comment:
This feels a bit final. Is that any error or an error from the dns provider?

Reply:
Any error: from provider failures to MZ not being in a ready state. We treat all the errors in the same way. To be honest, in reality, there is one case - if it tries to update the record and fails to write the status to the CR - when the error would make it reconcile immediately (and that is exactly what we want). In any other case, the error will be put in the `Ready - false` condition and we would obey the `noRequeueDuration`. The idea is that there is something wrong in your configuration and you should first fix that and it will trigger reconciliation again. 
We have the power, however, to manipulate this behaviour through the returned duration right now: we could decide that for "managed zone not ready" we want to requeue the record in ~15 minutes but for an unknown policy type we might want to stop trying.

Reply:
Yeah I am just not sure we should always assume misconfiguration. There could be a network problem for example that means you can't reach the provider. So in my mind setting ready to false and requeuing but for `RequeueDuration` in these cases might not be a bad thing. IE allowing for some external state to have been fixed (perhaps there was something wrong with the zone or the credential for example) and then trying again. I realise we are documenting how it works right now which is fine, but I think we should capture whether we want it to behave that way permanently

Reply:
I'll create an issue for this, or rather append it to #80